### ISSUE TITLE
Get gradient of preprocessing without forward passing each time 

### ISSUE THREAD

**Issue by Myrvixen** [keras-team/keras#12879]
I'm currently working on a regression problem where I need to preprocess the raw input data in a certain way (what's begin done is not relevant) before I can feed it to my neural network. I'm aware that preprocessing is usually done outside the model, but as I also need the gradients of the output of the network wrt. the raw input data (used in the loss function), I need the preprocessing to be included in the computational graph.

From what I've gathered, I need to feed the resulting tensor as input to the neural network in order for the two parts to be connected in the graph. An important note is that the preprocessing is done on the entire training set, meaning it only needs to be done once.

I've successfully implemented this using Keras backend functions, but the problem is that the preprocessing part seems to be performed at every forward pass of the model, which is both really unnecessary, and really slow. My hope was that the input tensor would be treated as some static (but differentiable) input to the network, with the forward-backward propagation simply flowing between the network input and output during training, but it doesn't seem to be working that way.

As my code is quite large and intricate, I'll leave a "semi-pseudic" code to simply show the gist of what I'm trying to do.

```python
import tensorflow.keras.backend as K
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Model

raw_data, target_data = load_data()
raw_tens = K.variable(raw_data)
out_tens = K.variable(target_data)

def preprocess(data_tens):
    input_tensor = ... # Preprocess the raw data using backend functions
    # Print statement to see when this operation is being performed
    input_tensor = K.print_tensor(input_tensor, message="Passed!")

    return input_tensor

in_tens = preprocess(raw_tens)

model_input = Input(tensor=in_tens)

# Define layers in the model
h = Dense(100, activation='tanh')(model_input)
...
output = Dense(1, activation='linear')(h)

# Define a custom loss function using the gradient wrt raw data
def custom_loss(y_true, y_pred):
    grad = K.gradients(y_pred, raw_tens)[0]

    loss = ... # some loss including the computed gradient
    return loss

# Define model and compile model using a custom loss function
model = Model(inputs=model_input, outputs=output)
model.compile(loss=custom_loss, optimizer='adam')

# Fit model
model.fit(y = out_tens, epochs = 1000, steps_per_epoch = 1)
```

In the preprocessing function I have added a print-statement to the tensor that prints whenever the function is called, and it does indeed print at every epoch during training. This also happens even if I don't include the gradient in the loss function, showing that it most likely prints in the forward pass. The model does seem to train and predict as intended however, so the implementation seems to be correct.

There might be some flaws in my logic/understanding of how the graph works, and I'm hoping there is a simple fix to my problem.

To summarize:

1.  How do I feed a "constant", differentiable tensor as input to a neural network, without the tensor being calculated anew during each forward pass through the network?

As the preprocessed tensor contains all training samples, I would also like to split and train on smaller batches, so:

2.  Given there exists a solution to 1., how does batch-training tie into this? Do I have to create a generator and fit with fit_generator, or is it possible to let Keras handle it with fit?

Hope the problem was clear enough, and I'm grateful for all help/leads!
